Create Audience Tile Error component.

benbowler opened the pull request:
## Summary

<!-- Please reference the issue this PR addresses in the following list. -->
Addresses issue:

- #8228 

## Relevant technical choices

See comment below.

## PR Author Checklist

- [x] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 5.2 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [ ] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [ ] Run the code.
- [ ] Ensure the acceptance criteria are satisfied.
- [ ] Reassess the implementation with the IB.
- [ ] Ensure no unrelated changes are included.
- [ ] Ensure CI checks pass.
- [ ] Check Storybook where applicable.
- [ ] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [ ] Ensure the PR has the correct target branch.
- [ ] Double-check that the PR is okay to be merged.
- [ ] Ensure the corresponding issue has a ZenHub release assigned.
- [ ] Add a changelog message to the issue.

hussain-t commented on assets/js/modules/analytics-4/components/dashboard/AudienceSegmentation/AudienceTile/AudienceTileError.js:
```diff
@@ -0,0 +1,160 @@
+									),
+									{
+										HelpLink: (
+											<Link
+												href="https://sitekit.withgoogle.com/documentation/troubleshooting/analytics/#insufficient-permissions"
+												external
+												hideExternalIndicator
+											>
+												{ __(
+													'Get help',
+													'google-site-kit'
+												) }
+											</Link>
+										),
+									}
+								) }
+							</p>
+						) }
+					</div>
+					<div className="googlesitekit-audience-segmentation-tile-error__cta">
+						{ hasInsufficientPermissionsError ? (
+							<Fragment>
+								<Button
+									href={ requestAccessURL }
+									target="_blank"
+									danger
+								>
+									{ __(
+										'Request access',
+										'google-site-kit'
+									) }
+								</Button>
```
Did you consider reusing the `ReportErrorActions` component? Please check the IB of [8230](https://github.com/google/site-kit-wp/issues/8230).

cc: @techanvil

techanvil replied:
Thanks @hussain-t. As we have that specced for #8230 and the changes and reuse of `ReportErrorActions` would be applicable here too, and would mean we can avoid copying part of `ReportErrorActions` here, that would be a good shout.

hussain-t commented on assets/js/modules/analytics-4/components/dashboard/AudienceSegmentation/AudienceTile/AudienceTileError.stories.js:
```diff
@@ -0,0 +1,116 @@
+
+function Template( { setupRegistry = async () => {}, viewContext, ...args } ) {
+	const setupRegistryCallback = async ( registry ) => {
+		provideModules( registry );
+		provideModuleRegistrations( registry );
+		await registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetSettings( {} );
```
These mocks seem unnecessary.

benbowler replied:
The receive settings mock was not needed but the module registration is needed to store and retrieve errors and their related reports.

hussain-t reviewed (changes requested):
Thanks, @benbowler. It's a good start. Please look at my comments. Also, please provide the GH storybook link in the QAB.

techanvil commented on assets/js/modules/analytics-4/components/dashboard/AudienceSegmentation/AudienceTile/AudienceTileError.js:
```diff
@@ -0,0 +1,160 @@
+						{ hasInsufficientPermissionsError && (
+							<p className="googlesitekit-audience-segmentation-tile-error__explanation">
+								{ __(
+									"You'll need to contact your administrator.",
+									'google-site-kit'
```
Thanks @hussain-t, that's a good point. I'd imagine we can simply increase the title size, but I've asked @sigal-teller on [Figma](https://www.figma.com/file/7pSrkEy8t00BcYRAi9LjjH?type=design&node-id=1396-25492&mode=design#748279313) to provide a design to clarify this one.

techanvil replied:
Hey @hussain-t @benbowler, Sigal has provided a [Figma design](https://www.figma.com/file/7pSrkEy8t00BcYRAi9LjjH/Audience-Segmentation?type=design&node-id=3395-16789&mode=design&t=xSRcYzYBc9jAMKpQ-0) for the "insufficient permissions" variant. Notably the descriptive text is smaller, using the `body/small` typography.

hussain-t replied:
Thanks, @techanvil

benbowler commented:
Hey @hussain-t + @techanvil, based on your comments I refactored this `AudienceTileError` component to use the `ReportErrorActions` component, to do that I had to add some additional props and code to it. It did greatly reduce the duplication of code into the `AudienceTileError` component.

A number of other elements from the original review were removed as part of this work and I addressed all the rest.

Back to you @hussain-t for another look.

hussain-t commented on assets/js/components/ReportErrorActions.js:
```diff
@@ -173,6 +196,12 @@ ReportErrorActions.propTypes = {
 	getHelpClassName: PropTypes.string,
 };
+
+ReportErrorActions.defaultProps = {
+	showGetHelpLink: true,
+};
```
`defaultProps` is deprecated in the latest React version, though we use it in our codebase. You can either pass a default value to the `showGetHelpLink` prop or rename it to `hideGetHelpLink` and use it accordingly, i.e., render the link if it's falsy.

hussain-t commented on assets/js/modules/analytics-4/components/dashboard/AudienceSegmentation/AudienceTile/AudienceTileError.js:
```diff
@@ -0,0 +1,100 @@
+			{
+				HelpLink: (
+					<Link
+						href="https://sitekit.withgoogle.com/documentation/troubleshooting/analytics/#insufficient-permissions"
+						external
```
We don't have to pass the `href` link here as it will be passed down to `GetHelpLink` as a prop from `ReportErrorActions`. See: https://github.com/google/site-kit-wp/blob/a849ab3424083af113a910e84b3e16e05c27dfce/assets/js/components/ReportErrorActions.js#L176

For example usage, see: 
https://github.com/google/site-kit-wp/blob/0a6fc5f62fe0da245904a8d8f10dbf58d9e8e90d/assets/js/components/KeyMetrics/GetHelpLink.js#L35-L47

benbowler replied:
I did originally implement it this way but I wasn't able to verify that resolve returned the correct support link from the proxy. Should I mock the `getErrorTroubleshootingLinkURL` selector to be able to test this or is this overkill for storybook?

hussain-t replied:
IMO, we can skip mocking for the storybook.

hussain-t commented on assets/js/modules/analytics-4/components/dashboard/AudienceSegmentation/AudienceTile/AudienceTileError.js:
```diff
@@ -0,0 +1,100 @@
+						<ReportErrorActions
+							moduleSlug="analytics-4"
+							error={ errors }
+							GetHelpLink={
+								hasInsufficientPermissionsError && GetHelpLink
+							}
```
The `GetHelpLink` prop expects an `elementType` or `undefined`. However, if the condition fails, it will be a boolean value.

<img width="1238" alt="Screenshot 2024-03-28 at 3 42 50 PM" src="https://github.com/google/site-kit-wp/assets/24408261/b482935f-dfc1-4bfc-9645-f3f81f21fc80">


```suggestion
GetHelpLink={
								hasInsufficientPermissionsError
									? GetHelpLink
									: undefined
							}
```

hussain-t commented on assets/sass/components/analytics-4/audience-segmentation/_googlesitekit-audience-segmentation-tile-error.scss:
```diff
@@ -0,0 +1,65 @@
+	.googlesitekit-audience-segmentation-tile-error {
+		background-color: $c-site-kit-sk-0;
+		border: 2px solid $c-red-r-500;
+		border-radius: 16px;
```
```suggestion
		border-radius: $br-sm;
```

hussain-t reviewed (changes requested):
Thanks, @benbowler. Looks great. However, we need to make a few more changes before we move it to MR. Please look at my comments.

hussain-t reviewed (approved):
Nice one, @benbowler. LGTM 🎖️

tofumatt commented on assets/js/modules/analytics-4/components/dashboard/AudienceSegmentation/AudienceTile/GetHelpLink.js:
```diff
@@ -0,0 +1,53 @@
+ */
+import Link from '../../../../../../components/Link';
+
+export default function GetHelpLink( { linkURL } ) {
+	return createInterpolateElement(
+		__(
+			'Contact your administrator. Trouble getting access? <HelpLink />',
+			'google-site-kit'
+		),
+		{
+			HelpLink: (
+				<Link href={ linkURL } external hideExternalIndicator>
+					{ __( 'Get help', 'google-site-kit' ) }
+				</Link>
+			),
+		}
+	);
+}
+
+GetHelpLink.propTypes = {
+	linkURL: PropTypes.string.isRequired,
+};
```
We already have an extremely similar component here: https://github.com/google/site-kit-wp/blob/e0232dd614844f7b3f3bd11a539e91821403bdb6/assets/js/components/KeyMetrics/GetHelpLink.js#L35 and to a lesser extent here: https://github.com/google/site-kit-wp/blob/e0232dd614844f7b3f3bd11a539e91821403bdb6/assets/js/components/setup/CompatibilityChecks/GetHelpLink.js#L48

I think we're better served extracting as much of these as possible to a common component rather than duplicating them like this a lot.

I'd also say that the component should really just be outputting the link, and leave the extra text/`createInterpolateElement`-ing to the `assets/js/modules/analytics-4/components/dashboard/AudienceSegmentation/AudienceTile/AudienceTileError.js` component 🤔

benbowler replied:
@tofumatt if you bring `createInterpolateElement` out why have a sperate component here in the first place? It could simply be an inline `Link` in the parent component. What do you think is best here, to create a single `GetHelp`.

I do think it's cleaner to have a separate component here otherwise you end up with `createInterpolateElement` inside of a ternary which is messy. 

[site-kit-wp/assets/js/components/KeyMetrics/GetHelpLink.js](https://github.com/google/site-kit-wp/blob/e0232dd614844f7b3f3bd11a539e91821403bdb6/assets/js/components/KeyMetrics/GetHelpLink.js#L35)

We could unify these components but they're quite simple and I don't see massive value in combining/reusing theses across the Key Metrics and Audience Segmentation, what do you think is the best?

tofumatt replied:
I agree actually, I don't really think the `GetHelp` link there is useful. The `assets/js/components/setup/CompatibilityChecks/GetHelpLink.js` component makes a bit more sense because it has a selector that fills in the URL, but the Key Metrics one isn't really doing anything.

If you're suggesting we just nix the `GetHelpLink` I'm fine with that, at least as a separate component. I guess just do whatever seems the most straightforward, but the `GetHelpLink` doesn't really seem like something worth making its own component over, so happy to remove it and just inline the content 🤔

benbowler replied:
@tofumatt actually on looking into this further I remembered why it must be a separate component. In the initial implementation I duplicated lots of logic from `ReportErrorActions`. In an update I refactored so that we could use the `ReportErrorActions` component itself. Because of this we must pass the help link as a component so that it can get the `linkUrl` passed as a prop:

https://github.com/google/site-kit-wp/blob/6acae9f04f9d9c6293372d640b292c92038b1ab4/assets/js/components/ReportErrorActions.js#L175-L177

I think combining the 3X `GetHelpLink` components feel unnecessary to me because these components are so simple on their own and combining them will strip them down to very little.

tofumatt commented on assets/sass/components/analytics-4/audience-segmentation/_googlesitekit-audience-segmentation-tile-error.scss:
```diff
@@ -0,0 +1,65 @@
+			display: flex;
+			flex-direction: column;
+			gap: $grid-gap-desktop;
+			margin: 34px auto;
```
Any reason for the `34px` spacing? Most of our stuff tries to follow an `8px` grid I thought, so this seems a bit "off" 🤔

tofumatt commented on assets/sass/components/analytics-4/audience-segmentation/_googlesitekit-audience-segmentation-tile-error.scss:
```diff
@@ -0,0 +1,65 @@
+			gap: $grid-gap-desktop;
+			margin: 34px auto;
+			text-align: center;
+			width: 352px;
```
Why the set width on this container? Should this be a max width instead?

tofumatt reviewed (changes requested):
Looks like there are some issues with the tile in an error state at very small screen sizes:

![Screenshot 2024-03-28 at 22 35 45](https://github.com/google/site-kit-wp/assets/90871/53163ce0-be9e-497b-bfe8-6e22e491b46c)
![Screenshot 2024-03-28 at 22 34 44](https://github.com/google/site-kit-wp/assets/90871/bdf0f98b-2fe3-4760-b16a-7a72bc0106a1)
![Screenshot 2024-03-28 at 22 34 38](https://github.com/google/site-kit-wp/assets/90871/c7a9572f-cfbe-4ba2-8e7e-ad6476e53591)

We tend to design for screens as small as `320px` wide, so here the error tiles fails while the other versions of this tile are fine. Even at iPhone SE sizes (`375px` wide), it's still quite tight with margin issues.

-----

Largely looks good, but a few comments/questions, and I think the small viewport sizing needs some adjustment.
